_posts/2024-11-05-lifetime-event-loop.md

```diff
@@ -0,0 +1,103 @@
+tags: [feature, eventloop, concurrency, 84codes]
+---
+
+A core component of Crystal's concurrency model is the **event loop**. It
+integrates asynchronous operations into the runtime, enabling other fibers to
+run while one fiber waits to read data on a socket, for example.
+
+We're changing how the event loop operates under the hood on Unix systems in [#14996].
+
+The new implementation integrates the Crystal event loop directly with system
+selectors, [`epoll`](https://linux.die.net/man/7/epoll) (Linux, Android) and
+[`kqueue`](https://man.freebsd.org/cgi/man.cgi?kqueue) (BSDs, macOS) instead of
+going through [`libevent`](https://libevent.org/).
+
+We're removing an external dependency and take control over a core runtime
+feature. It also changes how file descriptors are treated: instead of being
+added and removed on every blocking IO operation, the file descriptors are now
+added once and kept for their full lifetime, which is how epoll and kqueue have
+been designed. This reduces overhead and improves performance.
+
+This post highlights the relevant information for users.
+
+More technical details are available in [RFC #0009].
+
+## Effects
+
+The new implementation has been merged into `master` and is available in
+[nightly builds](/install/nightlies).
+
+No changes in user code are required, everything plugs right in.
+
+The new implementation is supported on Linux, macOS, FreeBSD and Android and
+automatically enabled on these systems. Read [more about
+availability][availability] in the RFC.
+
+For the time being other UNIX operating systems still use the `libevent` event
+loop by default (regressions, issues or untested). You can still force enable it
+using `-Devloop=kqueue` (BSDs) or `-Devloop=epoll` (e.g. Solaris).
+
+Windows is unaffected and keeps using `IOCP`.
+
+Dropping `libevent` removes an external runtime dependency from Crystal
+programs.
+
+## Caveats
+
+In some cases the new implementation may cause issues. You can switch
+back to the old event loop implementation with the compile-time flag
+`-Devloop=libevent`.
+
+We are aware of some potential regressions but believe they are quite rare and
+should not hinder general availability of this new feature. Exposure through
+nightly builds should help us gather more usage data to assess whether there are
+any more noteworthy implications we have not been aware of.
+
+### Timers and Timeouts
+
+At this point, there's a missing optimization for timers (`sleep 1.second`) and
+timeouts (`socket.read_timeout = 1.second`). In case you're using lots of them,
+it might lead to performance degradation.
+
+This is only a temporary limitation until we finish the implementation of an
+efficient data structure. We're already working on that.
+
+### Multi-Threading
+
+The new implementation works well with the multi-threading preview
+(`-Dpreview_mt`), having one event loop instance per thread.
+
+There is on caveat though: Moving file descriptors with pending operations
```

Comment:
As with the RFC, the breaking change should be shouted out. Also, the distinction between transferring FDs without pending operations should stand. More like "It is possible to transfer a file descriptor between threads **as long as there are no pending operations**."

Reply:
This last bit is to not make it sound like trasferring is not possible at all (for the casual reader)

Reply:
@straight-shoota have you seen this?